server/src/frameworks/Truffle/TruffleProject.ts

```diff
@@ -0,0 +1,245 @@
+      const resolvedSolcVersion = semver.maxSatisfying(
+        this.serverState.solcVersions,
+        configSolcVersion
+      );
+      if (resolvedSolcVersion === null) {
+        throw new Error(
+          `No version satisfies ${configSolcVersion}. Available versions are: ${this.serverState.solcVersions}`
+        );
+      }
+      this.resolvedSolcVersion = resolvedSolcVersion;
+
+      // Load contracts directory
+      this.sourcesPath = path.resolve(
+        this.basePath,
+        config?.contracts_directory ?? "contracts"
+      );
+      this.status = Status.INITIALIZED_SUCCESS;
+    } catch (error) {
+      const errorMessage = `Error loading config file ${this.configPath}: ${error}`;
+      this.status = Status.INITIALIZED_FAILURE;
+      this.initializeError = errorMessage;
+      throw new Error(errorMessage);
+    }
+  }
+
+  public async fileBelongs(file: string): Promise<FileBelongsResult> {
+    let belongs: boolean;
+    if (this.status === Status.INITIALIZED_SUCCESS) {
+      belongs = [this.sourcesPath, this.testsPath].some((dir) =>
+        directoryContains(dir, path.dirname(file))
+      );
+    } else {
+      // Claim ownership under all contracts on the base folder if initialization failed
+      belongs = directoryContains(this.basePath, file);
+    }
+
+    return {
+      belongs,
+      isLocal: true,
+    };
+  }
+
+  public async resolveImportPath(
+    file: string,
+    importPath: string
+  ): Promise<string | undefined> {
+    // Absolute path
+    if (path.isAbsolute(importPath)) {
+      return existsSync(importPath) ? importPath : undefined;
+    }
+
+    // Relative path
+    if (importPath.startsWith(".") || importPath.startsWith("..")) {
+      const resolvedPath = path.resolve(path.dirname(file), importPath);
+      return existsSync(resolvedPath) ? resolvedPath : undefined;
+    }
+
+    // Truffle direct imports
+    if (importPath.startsWith("truffle")) {
+      try {
+        return require.resolve(importPath.replace("truffle", "truffle/build"), {
+          paths: [this.basePath, this.globalNodeModulesPath],
+        });
+      } catch (error) {}
+    }
+
+    // Node modules direct imports (local and global)
+    try {
+      return require.resolve(importPath, {
+        paths: [this.basePath, this.globalNodeModulesPath],
+      });
+    } catch (error) {}
+
+    return undefined;
+  }
+
+  public async buildCompilation(
+    sourceUri: string,
+    openDocuments: OpenDocuments
+  ): Promise<CompilationDetails> {
+    // Ensure project is initialized
+    if (this.status !== Status.INITIALIZED_SUCCESS) {
+      const buildError: BuildInputError = {
+        _isBuildInputError: true,
+        fileSpecificErrors: {},
+        projectWideErrors: [
+          {
+            type: "general",
+            message: `Truffle project couldn't initialize correctly: ${this.initializeError}`,
+            source: "truffle",
+          },
+        ],
+      };
+      throw buildError;
```

Comment:
Should this not originate within the initialize method?

Reply:
The design I have in mind is as follows: on validation, we build the input to solc (framework- dependant), and then send it to solc. There are 3 types of errors: solc error, framework error on file (e.g. unresolvable import), and framework error on project (e.g. couldn't load config). The diagnostics and status item are handled differently depending on the type of error. For instance, framework project-level errors don't show as diagnostics but they show on the status item on files that belong to them.

During project initialization, if there's a failure e.g. loading the config, the error is stored in case compilation is attempted on a file belonging to such project.